### PR TITLE
Cmake prefix path

### PR DESCRIPTION
**Pull request by dcbaker**
This is analogous to pkg_config_path, but for cmake. I'd really like to get this in for 0.51.0 for mesa to use cmake for llvm finding, as it's much easier to use for cross compiling (ie, it's possible) than llvm-config is. If it doesn't make it in it's no the end of the world.

@mensinda, maybe you'd be interested in looking at this? I also wrote functionality for CMAKE_MODULE_PATH before I realized that only works for FindModule.cmake files and not ModuleConfig.cmake files. I'm also not sure what the best way to test this is?

### CODE REVIEW

File: mesonbuild/dependencies/base.py [hunk 1]
```diff
@@ -1133,6 +1133,14 @@ def search():
...
         if cm_path:
             cm_args += ['-DCMAKE_MODULE_PATH={}'.format(';'.join(cm_path))]
+
+        if environment.is_cross_build() and self.want_cross:
+            pref_path = ';'.join(self.env.coredata.builtins['cross_cmake_prefix_path'].value)
+        else:
+            pref_path = ';'.join(self.env.coredata.builtins['cmake_prefix_path'].value)
+        if pref_path:
+            cm_args.append('-DCMAKE_PREFIX_PATH={}'.format(pref_path))
+
```
Comment (ignatenkobrain) [inline mesonbuild/dependencies/base.py]: __[Flake8]__ [W293] blank line contains whitespace ([link](https://sider.review/gh/repos/19784232/pulls/5414#issue-4108700))

<sub>posted by Sider</sub>

**Comment by mensinda** [comment]
I think a unit test would be nice to have for this. The actual change looks good to me.

File: mesonbuild/dependencies/base.py [hunk 2]
```diff
@@ -1131,9 +1131,15 @@ def search():
...
-            cm_args += ['-DCMAKE_MODULE_PATH={}'.format(';'.join(cm_path))]
-        if not self._preliminary_find_check(name, cm_path, environment.machines[for_machine]):
+
+        if environment.is_cross_build() and self.want_cross:
+            pref_path = self.env.coredata.builtins['cross_cmake_prefix_path'].value
+        else:
+            pref_path = self.env.coredata.builtins['cmake_prefix_path'].value
+        if pref_path:
+            cm_args.append('-DCMAKE_PREFIX_PATH={}'.format(';'.join(pref_path)))
+
```
Comment (ignatenkobrain) [inline mesonbuild/dependencies/base.py]: __[Flake8]__ [W293] blank line contains whitespace ([link](https://sider.review/gh/repos/19784232/pulls/5414#issue-4108710))

<sub>posted by Sider</sub>

**Comment by dcbaker** [comment]
Now with tests, and fixes for the function that pre-fails the dependency check

File: mesonbuild/dependencies/base.py [hunk 3]
```diff
@@ -1131,9 +1131,15 @@ def search():
         cm_path = stringlistify(extract_as_list(kwargs, 'cmake_module_path'))
         cm_args = stringlistify(extract_as_list(kwargs, 'cmake_args'))
         cm_path = [x if os.path.isabs(x) else os.path.join(environment.get_source_dir(), x) for x in cm_path]
-        if cm_path:
```
Comment (mensinda) [inline mesonbuild/dependencies/base.py]: `CMAKE_MODULE_PATH` is still required for custom Find<Dep>.cmake files (see CI failures).

Comment (dcbaker) [reply mesonbuild/dependencies/base.py]: oops, that was not supposed to be in there. that's leftover from the original attempt I made to use CMAKE_MODULE_PATH. Fixed now.

**Pull request merged**